File: LeetCode/practice-cpp/LeetCode-1.cpp
// 两数之和
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        std::map<int, int> numsMap; // 使用 map 加快查询
        for (int i = 0; i < nums.size(); i++) {
            numsMap[nums[i]] = i;
        }
        
        for (int i = 0; i < nums.size(); i++) {
            auto index = numsMap.find(target - nums[i]);
            if (index != numsMap.end() && index->second != i) {
                return vector<int>{i, index->second};
            }
        }
        return vector<int>{};
    }
};

File: LeetCode/practice-rust/LeetCode-1.rs
// 两数之和
use std::collections::HashMap;

struct Solution;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut num_index_map: HashMap<i32, usize> = HashMap::new();
        for (index, &num) in nums.iter().enumerate() {
            let component = target - num;
            if let Some(&another) = num_index_map.get(&component) {
                return vec![index as i32, another as i32];
            }
            num_index_map.insert(num, index);
        }
        panic!("Cann't found result!");
    }
}

fn main () {
    let result = Solution::two_sum(vec![1, 2, 3, 4], 6);
    println!("{:?}", result);
}